governance/node.go

```diff
@@ -0,0 +1,233 @@
+		return
+	}
+
+	// get all the unarchived agreements
+	agreements, err := persistence.FindEstablishedAgreementsAllProtocols(w.db, policy.AllAgreementProtocols(), []persistence.EAFilter{persistence.UnarchivedEAFilter()})
+	if err != nil {
+		glog.Errorf(logString(fmt.Sprintf("Unable to retrieve all the  from the database, error %v", err)))
+		return
+	}
+
+	// cancel the agreement if needed
+	for _, ag := range agreements {
+		agreementId := ag.CurrentAgreementId
+		if ag.AgreementTerminatedTime != 0 && ag.AgreementForceTerminatedTime == 0 {
+			glog.V(3).Infof(logString(fmt.Sprintf("skip agreement %v, it is already terminating", agreementId)))
+		} else {
+			bCancel, err := w.agreementRequiresService(ag, svcSpecs)
+			if err != nil {
+				glog.Errorf(fmt.Sprintf("%v", err))
+			}
+
+			if bCancel {
+				glog.V(3).Infof(logString(fmt.Sprintf("ending the agreement: %v", agreementId)))
+
+				reason := w.producerPH[ag.AgreementProtocol].GetTerminationCode(producer.TERM_REASON_NODE_USERINPUT_CHANGED)
+
+				eventlog.LogAgreementEvent(
+					w.db,
+					persistence.SEVERITY_INFO,
+					persistence.NewMessageMeta(EL_GOV_START_TERM_AG_WITH_REASON, ag.RunningWorkload.URL, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason)),
+					persistence.EC_CANCEL_AGREEMENT,
+					ag)
+
+				w.cancelAgreement(agreementId, ag.AgreementProtocol, reason, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason))
+
+				// send the event to the container in case it has started the workloads.
+				w.Messages() <- events.NewGovernanceWorkloadCancelationMessage(events.AGREEMENT_ENDED, events.AG_TERMINATED, ag.AgreementProtocol, agreementId, ag.GetDeploymentConfig())
+				// clean up microservice instances if needed
+				w.handleMicroserviceInstForAgEnded(agreementId, false)
+			}
+		}
+	}
+}
+
+// Node pattern has been changes. Go unregister and re-register.
+func (w *GovernanceWorker) handleNodeExchPatternChanged(shutdown bool, new_pattern string) {
+	glog.V(5).Infof(logString(fmt.Sprintf("handling node pattern changes")))
+
+	pDevice, err := persistence.FindExchangeDevice(w.db)
+	if err != nil {
+		glog.Errorf(logString(fmt.Sprintf("error getting device from the local database. %v", err)))
+		return
+	}
+
+	if shutdown {
+		// unconfigure the node only if it is not already unconfiguring
+		if pDevice.Config.State != persistence.CONFIGSTATE_UNCONFIGURING {
+			// set the device config stat to unconfiguring
+			_, err = pDevice.SetConfigstate(w.db, pDevice.Id, persistence.CONFIGSTATE_UNCONFIGURING)
+			if err != nil {
+				glog.Errorf(logString(fmt.Sprintf("error persisting unconfiguring on node object: %v", err)))
+				return
+			}
+			// set the node shutdown message
+			w.Messages() <- events.NewNodeShutdownMessage(events.START_UNCONFIGURE, false, false)
+		}
+	} else {
+		// the device is up again and rereg the device with the new pattern
+		if err := w.changeNodePattern(pDevice, new_pattern); err != nil {
```

Comment:
What state is the device in if this error occurs? Is it unregistered? How does the user recover?

Reply:
configured. First the pattern has been verified before shutdown.

Reply:
If anything is wrong, user can set the pattern back on the exchange.